src/log.c

```diff
@@ -36,70 +36,378 @@
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
 
+#include <rdma/fi_errno.h>
 #include "fi.h"
 #include "fi_log.h"
 
+#define ARRAY_SIZE(x) (sizeof((x)) / sizeof((x)[0]))
+
+static struct slist provider_handles;
+static struct log_providers provider_list;
+
+const char *subsystems[] = {
+	[FABRIC] = "fabric",
+	[DOMAIN] = "domain",
+	[EP_CM] = "ep_cm",
+	[EP_DM] = "ep_dm",
+	[AV] = "av",
+	[CQ] = "cq",
+	[EQ] = "eq",
+	[MR] = "mr"
+};
+
+const char *log_levels[] = {
+	[LOG_WARN] = "warn",
+	[LOG_TRACE] = "trace",
+	[LOG_INFO] = "info"
+};
+
+/*
+ * By default support logging from all providers and subsystems. This will
+ * change if the relevant environment variables are set.
+ */
+uint64_t log_mask = PROV_MASK | SUBSYS_MASK;
+
+typedef void found_func_t(char *ptr);
+
 /* General implementation note: these functions currently use multiple fprintfs
  * in a row, which can render in an ugly fashion for multithreaded code and for
  * some mpirun implementations.  If this bugs anyone enough then we can convert
  * them to snprintf to build up the printout in a single buffer.
  */
 
-int fi_log_level = INT_MIN;
+static int compare_names(struct slist_entry *item, const void *arg)
+{
+	struct provider_parameter *target =
+	    container_of(item, struct provider_parameter, entry);
+	return !strcmp(target->name, arg);
+}
+
+static bool check_negated(char **str)
+{
+	bool negated = false;
+
+	if (**str == '^') {
+		(*str)++;
+		negated = true;
+	}
+
+	return negated;
+}
+
+static char *alloc_env(char *name)
+{
+	char *env = getenv(name);
+	char *ret = NULL;
+
+	if (!env) {
+		return ret;
+	}
+
+	ret = strdup(env);
+	if (!ret) {
+		fprintf(stderr, "%s: %s", __func__, strerror(errno));
+	}
+
+	return ret;
+}
+
+struct prov_handle *fi_log_register_provider(const char *prov_name)
+{
+	struct prov_handle *handle;
+	struct slist_entry *found;
+	struct provider_parameter *item;
+
+	static char curr_pos = 1;
+
+	handle = calloc(sizeof(*handle), 1);
+	if (!handle) {
+		goto err;
+	}
+
+	handle->name = strdup(prov_name);
+	if (!handle->name) {
+		goto cleanup;
+	}
+
+	handle->bitpos = curr_pos;
+
+	curr_pos <<= 1;
+
+	found = slist_remove_first_match(&provider_list.names, compare_names,
+					 handle->name);
+
+	/*
+	 * Since providers are loaded dynamically we can't just negate the mask
+	 * at the end after all providers in the negated list have been added.
+	 * Need to maintain a list of providers that can print at any given
+	 * time before all of the providers have loaded.
+	 */
+	if (found) {
+		if (provider_list.negated) {
+			log_mask &= ~EXPAND(handle->bitpos, PROV_OFFSET);
+		} else {
+			log_mask |= EXPAND(handle->bitpos, PROV_OFFSET);
+		}
+
+		item = container_of(found, struct provider_parameter, entry);
+		free(item->name);
+		free(item);
+	}
+
+	slist_insert_tail(&handle->entry, &provider_handles);
```

Comment:
@shefty My reasoning here might be faulty so I'll run through my thoughts so you could understand the decisions I made: 

There are two lists: the providers gathered from the environment variable and the list of handles. The first list is required because I'm receiving the provider list from the environment variable before any providers are loaded. This is necessary to allow use of the logging functions inside of the INI functions. The second list contains the handles that contain the provider name and its associated bit position. Storing the name in the structure is mainly so I don't have to do some lookup when printing the provider name inside of the logging functions. 

How would you propose I do the provider list? I thought about it for a while and didn't really know where to place it so that it worked the best.